sources/packages/api/src/services/application/application.service.ts

```diff
@@ -950,4 +955,52 @@ export class ApplicationService extends RecordDataModelService<Application> {
       ) <= COE_WINDOW
     );
   }
+  /**
+   * Deny the Confirmation Of Enrollment(COE) for an Application.
+   * Updates only applications that have the COE status as required.
+   * @param applicationId application id to be updated.
+   * @param locationId location that is setting the offering.
+   * @param coeDeniedReasonId COE Denied reason id for a student application.
+   * @param otherReasonDesc when other is selected as a COE denied reason, text for the reason
+   * is populated.
+   * @returns updated application.
+   */
+  async setDeniedReasonForConfirmationOfEnrollment(
+    applicationId: number,
+    locationId: number,
+    coeDeniedReasonId: number,
+    otherReasonDesc?: string,
+  ): Promise<Application> {
+    const application = await this.repo.findOne({
+      id: applicationId,
+      location: { id: locationId },
+      coeStatus: COEStatus.required,
+      applicationStatus: Not(
+        In([
+          ApplicationStatus.completed,
+          ApplicationStatus.overwritten,
+          ApplicationStatus.cancelled,
+        ]),
+      ),
+    });
+    if (!application) {
+      throw new CustomNamedError(
+        "Not able to find an application that requires a COE to be completed.",
+        COE_REQUEST_NOT_FOUND_ERROR,
```

Comment:
I would suggest using the generic existing constant error INVALID_OPERATION_IN_THE_CURRENT_STATUS that, alongside the message will give the right error context and will allow the proper HTTP staus to be returned from the controller consuming this method.